### ISSUE TITLE
mAP drop in yolov5s with v2.0 release 

### ISSUE THREAD

**Issue by hal-314** [ultralytics/yolov5#496]
## ❔Question
First of all, you are doing an excellent job!! Keep doing it :D 

I notice that all yolov5 model have increase mAP with v2.0 release except Yolov5s that decreased it's mAP from 36.6 to 36.1 while maintaining the same GPU speed and number of parameters. Also, I see that Yolov5m accuracy is almost exactly the same than before. 

So, changes in 2.0 release have focused in the big yolo models? Do you plan increase yolov5s accuracy back to where it was with 1.0 release?

Thank you.

**Comment by glenn-jocher** [comment]
@hal-314 yes this is correct, very observant! The model architectures are identical to before, we have modified their construction a bit in yolo.py and also modified training to better balance output losses among the 3 layers. This resulted in the mAP changes, which appear to benefit the larger models at the expense of the smallest model as you noticed. 

We also trained 2 different architectures which greatly improved the smaller models but also slowed them down, so we neglected to release those two architectures. YOLOv5s should be trainable to the same mAP as before, or better with the default settings, or possibly by increasing the hyp obj by 20-30%. Smaller models overtrain less than larger ones.

**Comment by AlexWang1900** [comment]
@glenn-jocher  Hi~~!
May I ask how to decide the layer balance ? it seems obj loss 3x larger in new settings on Wheat detection dataset.
and box loss and obj loss ratio? should I weighted them to make them almost same?

**Comment by glenn-jocher** [comment]
@AlexWang1900 oh interesting. The new balance is based on empirical results on COCO. I measured the obj loss across the 3 layers for COCO and balanced them using the new balance terms in compute_loss(). Only obj losses are balanced due to their unique nature. The small object output layer is weighted 4X now compared to before, so most wheat objects must be matched in the small obj layer if you are seeing a 3X increase in obj loss. 

Every dataset is different though, so I can't actually predict if v2.0 will perform better or worse on wheat. On COCO YOLOv5x mAP improved from 48.4 to 49.0 when trained with the new v2.0 updates (no architecture changes).

**Comment by glenn-jocher** [comment]
@AlexWang1900 you might want to try to rebalance them yourself. The guideline I use is to look for the loss component that overfits first, which is usually obj. Then sometimes reducing obj helps delay overfitting when training from scratch, when training from pretrained weights I'm not so sure, I have less experience there. If you are seeing 3X loss, and obj is now 3X larger than GIoU, you may want to try an experiment:

- Train with hyp['obj'] = 1.00 (default v2.0)
- Train with hyp['obj'] = 0.67
- Train with hyp['obj'] = 0.33

**Comment by AlexWang1900** [comment]
Thanks a lot!!!! great guidelines!!!!



---Original---

**Comment by glenn-jocher** [comment]
@AlexWang1900 you're welcome! 

The strange thing is my main challenge with YOLOv5 is not really training to high mAP, it's preventing overtraining. Overtraining on the larger models in particular is terrible, I'm not entirely sure what to do. This is the old and new YOLOv5x. Default training is 300 epochs, from scratch, but overtraining begins to occur as early as 200 epochs.

You can see GIoU and cls bottom out at 250 epochs, but obj bottoms at 200 epochs. If I could delay it 50 more epochs so all 3 loss components bottom out at the same time, the result would improve greatly, but I'm struggling with exactly how to do that.

![results](https://user-images.githubusercontent.com/26833433/88448881-07d16100-cdf7-11ea-91d2-ff8ffb34449a.png)

**Comment by AlexWang1900** [comment]
from my team's work on Wheat and other classification&nbsp; here are some thoughts:
the bn in torch has a tracker,which doesn't exist in Tensorflow. that makes the bn mean and variance converge to the whole dataset mean and variance. especially for large epoches.&nbsp;
that may explain pytorch effdet can't out perform Tf official effdet map score.&nbsp;


we believe it's the bn parameters lack of variance as epoches gets larger,causes overfitting.&nbsp;
So we did these to overcome overfitting
1)mixup. for an easy approach,we just mix with background pictures at 50%chance. without changing labels. that brings large variance to bn I think.&nbsp;
2)cutout&nbsp;
3) disable trackers in torch bn module but we haven't tried it yet.&nbsp;
without mixup it overfits under 40epoches,now it can train more then80 with all validation losses still go down



---Original---

**Comment by glenn-jocher** [comment]
@AlexWang1900 that's interesting. mixup 50% of the images with the same ratio or a beta distribution ratio (with alpha=0.3?). Not having to update the labels is nice. I will try this on COCO then.

About the BN I would assume that as the statistics smooth during later epochs the result would get better, unless the BN stats are not representative of the val images (i.e. due to augmentations perhaps). You can test the model in .train() mode instead of .eval() mode for the BN layers to ignore the stats and use realtime mean and sigma for normalization, but this should always produce worse results I believe.

**Comment by glenn-jocher** [comment]
I modified a beta distribution for this by skewing the alpha params to one side to obtain ratios for this idea. Perhaps this is simply trending towards an exp distribution by doing this.
```python
                import numpy as np
                import matplotlib.pyplot as plt
                plt.hist(np.random.beta(.3, 1.0, 1000000), 300)
```
![Figure_1](https://user-images.githubusercontent.com/26833433/88465472-1c097280-ce78-11ea-8605-a59f3cd630af.png)

An alternative would be to use the same alphas np.random.beta(0.3, 0.3), and to flip the labels when the ratio exceeds 50%. What do you think?
```python
            if random.random() < 0.5:
                img2, labels2 = load_mosaic(self, random.randint(0, len(self.labels) - 1))
                r = np.random.beta(0.3, 0.3)  # mixup ratio, alpha=beta=0.3
                img = (img * r + img2 * (1 - r)).astype(np.uint8)
                labels = labels2 if r < 0.5 else labels
```
![Figure_1](https://user-images.githubusercontent.com/26833433/88465549-cf726700-ce78-11ea-9533-6f571ea49e66.png)

**Comment by AlexWang1900** [comment]
we&nbsp; fixed 0.5:0.5&nbsp; with external pictures.as for in domain pictures I think a normal distribution may be good? concentrated in 0.3:07--0.7:0.3,with both labels on
&nbsp;the beta distribution from original paper is for classification,which can use weighted label add,using same weghts as pictures mixup weights. but boxes can only be hard labels.&nbsp;
and even for classification,we found a uniform distribution is better. I think under some conditions beta distribution (a=b=1) is a flat line. the 8:2or 2:8&nbsp; is&nbsp;conservative,not to make too hard cases for model to learn.&nbsp;





---Original---

**Comment by glenn-jocher** [comment]
@AlexWang1900 yes, beta distribution is very interesting. alpha=1,1 is a uniform as you say, I think alpha=10,1 is a simple line with slope 1. I started training COCO YOLOv5l with alpha=20,4, which gives me this distribution, which I like because it rarely crosses 0.5, and doesn't waste time at 1.0 either. I don't modify the labels at all. This will take a long time to train, about a week.

I had another idea about introducing perspective transforms into the augmentation. I think rotation may be partly hurting mAP on COCO due to the rotated image edges during training, which never happens during testing. So for this I applied extra effort to applying a perspective transform about the mosaic origin, which retains orthogonal lines as in the zero rotation case:

Example 1
![Figure_4](https://user-images.githubusercontent.com/26833433/88467831-16b82200-ce90-11ea-8e79-49bbb27fafe2.png)

Example 2
![Figure_5](https://user-images.githubusercontent.com/26833433/88467840-2d5e7900-ce90-11ea-8234-aba198c0d1e0.png)

**Comment by AlexWang1900** [comment]
I don't know in object detection it took so longer for mixup. classification task must be much easier.
as for edges a random&nbsp; center crop may eliminate edges.&nbsp;
perspective change I haven't seen it in any&nbsp; paper. it's a new idea!!!
also a gridmask is the new best augmentation in classification. maybe it can be used in object detection.&nbsp;







---Original---

**Comment by AlexWang1900** [comment]
about bn. there must be a little distribution shift between train and Val and test. to overfit the val/test distribution there must be some random shift also in training process. like a random search. while pytorch bn trakcer makes that impossible,as epoch grows every batch update makes fewer and fewer contribution in bn.&nbsp;
most augmentations will&nbsp; just mean out the changes. doesn't change the distribution especially with bn tracker.&nbsp;
bn trakcer may makes training more stable by representing trainset distribution. but some small disturbance each batch brings the chance to overfit to val/test set.&nbsp;


as for open bn during validation. many efficientDet does that and have a huge gain. I think you saw worse result because batch size =1,during testing,that unstable bn mean&nbsp; too much.







---Original---

**Comment by ZeKunZhang1998** [comment]
> @AlexWang1900 yes, beta distribution is very interesting. alpha=1,1 is a uniform as you say, I think alpha=10,1 is a simple line with slope 1. I started training COCO YOLOv5l with alpha=20,4, which gives me this distribution, which I like because it rarely crosses 0.5, and doesn't waste time at 1.0 either. I don't modify the labels at all. This will take a long time to train, about a week.
> 
> I had another idea about introducing perspective transforms into the augmentation. I think rotation may be partly hurting mAP on COCO due to the rotated image edges during training, which never happens during testing. So for this I applied extra effort to applying a perspective transform about the mosaic origin, which retains orthogonal lines as in the zero rotation case:
> 
> Example 1
> ![Figure_4](https://user-images.githubusercontent.com/26833433/88467831-16b82200-ce90-11ea-8e79-49bbb27fafe2.png)
> 
> Example 2
> ![Figure_5](https://user-images.githubusercontent.com/26833433/88467840-2d5e7900-ce90-11ea-8234-aba198c0d1e0.png)

I think after rotate, sometime the box becomes not the best one to fit the object.

**Comment by glenn-jocher** [comment]
@ZeKunZhang1998 yes, the boxes are augmented using the same transforms as the images. This may result in suboptimal boxes at extreme augmentations (i.e. 45 degree rotations).
